### PR TITLE
[rtl] Remove stall cycle with BT ALU

### PR DESCRIPTION
**Pull request by tomroberts-lowrisc**
- Branches don't need to stall with the branch target ALU

Signed-off-by: Tom Roberts <tomroberts@lowrisc.org>

### CODE REVIEW

**Comment by tomroberts-lowrisc** [comment]
@GregAC apologies if you were already looking at this, but it overlapped with my fixed execution time work.

**Comment by GregAC** [comment]
I had left this in intentionally. After any jump or branch there is a stall due to a pipeline bubble. So for a jump in particular it still counts as a multi-cycle instruction, one cycle executing the jump, one or more cycles waiting for the pipeline bubbles. With the branch target ALU a taken branch should do exactly the same as an unconditional jump from a stalling perspective

This is a bit weird but aimed to keep the behaviour of the existing RTL as much as possible (which treated the pipeline bubble after the jump/branch as a stall).

Looking at it again I don't think we've actually got the same behaviour between jump/branch anyway, without BT ALU a branch will set stall_branch the first cycle, then in `MULTI_CYCLE` perform the pc change then go back to `FIRST_CYCLE`, whilst a jump will perform the pc change then go to `MULTI_CYCLE` possibly only staying there a single cycle regardless of how long the iside takes to perform the jump.

So in conclusion happy with this change but I think we need to alter the jump behaviour as well, probably just remove the `jump_in_dec` case from the state machine entirely and remove the `stall_jump` signal as I'm not sure it's doing anything useful.

Out of interest did you see any actual issues because of this or did you just see it and thought it looked wrong?

**Comment by tomroberts-lowrisc** [comment]
Ah I see that makes sense. No issue found, just thought it looked weird.

To my mind the stall due to the pipeline bubble is separate from the instruction. If we add branch prediction, that stall might not be there at all so probably makes sense to remove it from the instruction.

The jumps stall to allow the writeback to happen for JAL (first cycle sets the jump, second calculates and stores PC+4). It makes sense for the jump to do it that way around since it's more performant - the writeback second cycle happens in the pipeline bubble so has no cost.

I have a rough version of the RTL where I extend the BT ALU to work for jumps too, after which the second cycle isn't required since the writeback can happen in parallel.

**Comment by GregAC** [comment]
All makes sense, had forgotten about the need for writeback for JAL.

Had actually intended to extend the BT ALU for jumps, looks like I forgot, thanks for looking at sorting it out.

**Comment by GregAC** [comment]
In case it was uncertain - This all LGTM following the discussion (i.e. this RTL change in isolation without the jump changes I initially proposed).
